### ISSUE TITLE
Support references to external types

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi-go-provider#16]
Pulumi schema can contain references to externally defined types from other packages.  This does not yet appear to be supported by this library.  In principal it seems like it could be supported either automatically (if package name and version can be identified?) or via some manual mapping information users provide?

This is to support something like this:

```json
"properties": {
                "functions": {
                    "type": "array",
                    "items": {
                        "$ref": "/aws/v4.0.0/schema.json#/resources/aws:lambda%2Ffunction:Function"
                    },
                    "description": "The functions managed by this service."
                },
                "url": {
                    "type": "string",
                    "description": "The API Gateway REST API managed by this service, if any of the functions expose HTTP handlers."
                }
            },
```

Via something like this:

```go
type Service struct {
	pulumi.ResourceState

	// Outputs
	Url     pulumi.StringPtrOutput `pulumi:"url" provider:"output"`
	Lambdas []*lambda.Function     `pulumi:"lambdas" provider:"output"`

	// Inputs
	Functions FunctionMapInput     `pulumi:"functions"`
	Provider  ServiceProviderInput `pulumi:"provider"`
}
```

**Comment by iwahbe** [comment]
We can already do this in theory for types like `ServiceProviderOutput`. We can implement support for this generically (input+output) by traversing the `To*Output` of `FunctionMapInput` and then calling `ElementType` on an empty instance of that. This should be a transparent experience to users.

**Comment by lukehoban** [comment]
#21 definitely addressed other issues (I made more progress on #17 👏 ) - but didn't address the issue described in this issue.

For two reasons I believe:
1. The code noted above (and commented out in #17) triggers `panic: Detected pointer type during serialization - did you forget to dereference?`.
2. I'm fairly sure even if that was fixed or worked around, this would run into what's captured in #20?

If (2) is the only issue, this could be closed as a dupe of #20, but probably worth reopening this for now?  (I can't as I'm not a contributor on this repo :smile:).

**Comment by harribo11** [comment]
I believe this issue is that `[]*lambda.Function` is an array of pointers, not that it's an external type. I've opened a pull request (#23 ) which should address this.

**Comment by lukehoban** [comment]
Fixed with #23.
